### PR TITLE
Feat/Hyprland: Allow matching against Window Titles for Rewrite

### LINKED ISSUE #2561: Feature request: in hyprland/workspaces allow also for window title and regex
I think that the recent commit on Hyprland dynamic window names on workspaces (58e506a6754fae213e409bcd079d38d03f70b958) by @Syndelis was very cool. It works very well in my setup.  If I understood correctly, at the moment it is only possible to use the window class to set the icon. Would it be possible/easy to add also the window title and/or allow for regex strings? I am sorry that I cannot help with this as I am not really proficient in coding.

### PR DESCRIPTION
**Pull request by Syndelis**
Resolves #2561 

## About this PR

This PR builds upon #2529 to introduce matching against a window's title in order to get its icon. In order to do that, the function that matches against regexes will now be supplied a string in this format: `"class<WINDOW_CLASS> title<WINDOW_TITLE>"`, as opposed to just the class. This way, users may provide rules in the following forms:

```json
"window-rewrite": {
  "title<.*youtube.*>": "", // Windows whose titles contain "youtube"
  "class<firefox>": "", // Windows whose classes are "firefox"
  "class<firefox> title<.*github.*>": "", // Windows whose class is "firefox" and title contains "github". Note that "class" always comes first.
  "foot": "", // Windows that contain "foot" in either class or title. For optimization reasons, it will only match against a title if at least one other window explicitly matches against a title.
}
```

Rules are now also prioritized, from more specilized to least. In the previous example, the priority order would be:
1. GitHub — Contains both class and title matching;
2. YouTube — Matches against a title, which is preferred over simply a class;
3. FireFox — Explicitly matches against a class;
4. Foot — Doesn't explicitly match against anything and, thus, is considered low priority. This prevents having your browser be displayed as a terminal if you search for "foot" in FireFox, for example.

## Backwards Compatibility

Because of the possible problem pointed out in **4.**, matching against titles is disabled **unless** at least one rule explicitly matches against titles. If no rules do so, titles are not sent to the aforementioned function, and, instead, the string `"class<WINDOW_CLASS>"` is used.

This does mean, however, that rules written as `"^something_here$"` will not work anymore, and should be replace by either `"somethind_here"`, `"class<something_here>"` or `"^class<something_here>$"`. While I don't believe anyone might've written such a specific rule, I'm open to hear the reasoning for that and work out something that doesn't break this niche use case.

## Implementation Details

A refactor has been made to separate the logic behind the regex map, its cache and default values. This whole logic is now available as the `waybar::util::RegexCollection` class, which can be constructed from a `Json::Value`. This class also provides a way to sort rules, giving priority to whichever rule scores higher in the parameter function.

When titles are used in any rules, the workspace module will be subscribed to an additional event called `windowtitle`, which, [according to Hyprland's IPC documentation](https://wiki.hyprland.org/IPC/), only features the window's address as a parameter. This means that an additional call to Hyprland's `"clients"` socket1 must be made in order to retrieve the window's latest name.

The module is not subscribed to this event when titles aren't explicitly used simply because it would be of no use and could break compatibility as mentioned above.

### CODE REVIEW

File: src/modules/disk.cpp
Comment (Syndelis) [inline src/modules/disk.cpp:1]: This file is unrelated to this PR and was simply linted it in f8340d8

**Comment by khaneliman** [comment]
Looks like freebsd check failed to build properly.

**Comment by Syndelis** [comment]
> Looks like freebsd check failed to build properly.

Yeah, my last PR #2568 did too. Unfortunately I can't do much to help it; the logs don't mention the reason it failed, it just goes from "Copying files" to "/bin/bash failed". Would need to re-run it with debug logging enable, but I don't have permissions for that.

**Comment by khaneliman** [comment]
In file included from /usr/include/c++/v1/__functional/function.h:20:
  /usr/include/c++/v1/__memory/allocator_traits.h:298:9: error: no matching function for call to 'construct_at'
          _VSTD::construct_at(__p, _VSTD::forward<_Args>(__args)...);
          ^~~~~~~~~~~~~~~~~~~
  /usr/include/c++/v1/__config:826:15: note: expanded from macro '_VSTD'
  #define _VSTD std
                ^
  /usr/include/c++/v1/vector:778:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<waybar::util::Rule>>::construct<waybar::util::Rule, const std::regex &, std::string, int &, void, void>' requested here
      __alloc_traits::construct(this->__alloc(), _VSTD::__to_address(__tx.__pos_),
                      ^
  /usr/include/c++/v1/vector:1577:9: note: in instantiation of function template specialization 'std::vector<waybar::util::Rule>::__construct_one_at_end<const std::regex &, std::string, int &>' requested here
          __construct_one_at_end(_VSTD::forward<_Args>(__args)...);
          ^
  ../src/util/regex_collection.cpp:24:15: note: in instantiation of function template specialization 'std::vector<waybar::util::Rule>::emplace_back<const std::regex &, std::string, int &>' requested here
          rules.emplace_back(rule, it->asString(), priority);
                ^
  /usr/include/c++/v1/__memory/construct_at.h:36:16: note: candidate template ignored: substitution failure [with _Tp = waybar::util::Rule, _Args = <const std::regex &, std::string, int &>]: no matching constructor for initialization of 'waybar::util::Rule'
  constexpr _Tp* construct_at(_Tp* __location, _Args&& ...__args) {
                 ^
  1 error generated.

**Comment by Syndelis** [comment]
> This is what I see looking at the checks logs under` Run 'run' in VM`
> 
> ```
>   In file included from /usr/include/c++/v1/__functional/function.h:20:
>   /usr/include/c++/v1/__memory/allocator_traits.h:298:9: error: no matching function for call to 'construct_at'
>           _VSTD::construct_at(__p, _VSTD::forward<_Args>(__args)...);
>           ^~~~~~~~~~~~~~~~~~~
>   /usr/include/c++/v1/__config:826:15: note: expanded from macro '_VSTD'
>   #define _VSTD std
>                 ^
>   /usr/include/c++/v1/vector:778:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<waybar::util::Rule>>::construct<waybar::util::Rule, const std::regex &, std::string, int &, void, void>' requested here
>       __alloc_traits::construct(this->__alloc(), _VSTD::__to_address(__tx.__pos_),
>                       ^
>   /usr/include/c++/v1/vector:1577:9: note: in instantiation of function template specialization 'std::vector<waybar::util::Rule>::__construct_one_at_end<const std::regex &, std::string, int &>' requested here
>           __construct_one_at_end(_VSTD::forward<_Args>(__args)...);
>           ^
>   ../src/util/regex_collection.cpp:24:15: note: in instantiation of function template specialization 'std::vector<waybar::util::Rule>::emplace_back<const std::regex &, std::string, int &>' requested here
>           rules.emplace_back(rule, it->asString(), priority);
>                 ^
>   /usr/include/c++/v1/__memory/construct_at.h:36:16: note: candidate template ignored: substitution failure [with _Tp = waybar::util::Rule, _Args = <const std::regex &, std::string, int &>]: no matching constructor for initialization of 'waybar::util::Rule'
>   constexpr _Tp* construct_at(_Tp* __location, _Args&& ...__args) {
>                  ^
>   1 error generated.
> ```

My bad, I was looking at the wrong step. I'll take a look at it

**Comment by Syndelis** [comment]
Apparently the pipeline runs with Clang 14, which still didn't support [direct initialization](https://en.cppreference.com/w/cpp/feature_test#cpp_aggregate_paren_init)

**Comment by khaneliman** [comment]
Tested locally seems to work, as intended. @Alexays should be good to go

**Review by khaneliman — APPROVED** [review]

**Comment by Alexays** [comment]
Thanks!
